[OpenMP] Generate implicit default mapper for mapping array section.

jyu2-git opened the pull request:
This is only for struct containing nested structs with user defined mappers.

Add four functions:
1>buildImplicitMap: build map for default mapper
2>buildImplicitMapper:  build default mapper.
3>hasUserDefinedMapper for given mapper name and mapper type, lookup user defined map, if found one return true.
4>isImplicitMapperNeeded check if Mapper is needed

During create map, in checkMappableExpressionList, call isImplicitMapperNeeded when it return true, call buildImplicitMapper to generate implicit mapper and added to map clause.

jyu2-git committed:
Generate implicit default mapper for mapping array section.

This is only for struct containing nested structs with custom mappers.

Add three functions:
1>buildImplicitMap: build map for default mapper
2>buildImplicitMapper:  build default mapper.
3 processImplicitMapperWithMaps: go throuth each map clause create
  mapper as needed.

In processImplicitMapsWithDefaultMappers, when nested user defined
mapper is found, create a clause list (ClausesNeedImplicitMapper)
to generat mapper for the corespoing map clause.

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -5753,6 +5753,155 @@ static ExprResult buildUserDefinedMapperRef(Sema &SemaRef, Scope *S,
+  DeclarationName MapperId;
+  auto &DeclNames = Ctx.DeclarationNames;
+  MapperId = DeclNames.getIdentifier(&Ctx.Idents.get("default"));
+  OMPDeclareMapperDecl *DMD = OMPDeclareMapperDecl::Create(
```
auto *DMD

jyu2-git replied:
Thanks.  Changed.

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -5753,6 +5753,155 @@ static ExprResult buildUserDefinedMapperRef(Sema &SemaRef, Scope *S,
+    // declare mapper.
+    return;
+
+  for (int Cnt = 0, EndCnt = Clauses.size(); Cnt < EndCnt; ++Cnt) {
```
`for (int Cnt : seq<int>(Clauses.size()))`

jyu2-git replied:
Thanks Alexey.  Good to know.  Changed.

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -5753,6 +5753,155 @@ static ExprResult buildUserDefinedMapperRef(Sema &SemaRef, Scope *S,
+      buildImplicitMapper(S, BaseType, Stack, UDMapperRefs);
+    }
+    if (!UDMapperRefs.empty()) {
+      assert(UDMapperRefs.size() == C->varlist_size());
```
Assertion message

jyu2-git replied:
Added.  Thanks.

jyu2-git commented on clang/include/clang/AST/OpenMPClause.h:
```diff
@@ -5556,6 +5556,7 @@ class OMPMappableExprListClause : public OMPVarListClause<T>,
     MapperIdInfo = MapperId;
   }
 
+public:
```
But I need to call this function after I create new mapper?  Do you mean there is other way to call this function?
Thanks.

jyu2-git replied:
Thanks Alexey!  I changed!  Instead generate implicit mapper during map creation.

jyu2-git committed:
Thanks Alexey for the review.  This is address his comments.

Rename the tests.

jyu2-git committed:
Fix format.

jyu2-git committed:
Thanks Alexey for the review.

This is generate implicit mapper when map is created instead after map
created.

jyu2-git committed:
Remove unused code.

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -20796,6 +20796,150 @@ struct MappableVarListInfo {
+  // Dummy variable _s for Mapper.
+  ImplicitName.setName(
+      Ctx.DeclarationNames.getIdentifier(&Ctx.Idents.get("_s")));
+  DeclarationName VN = ImplicitName.getName();
+  TypeSourceInfo *TInfo =
+      Ctx.getTrivialTypeSourceInfo(BaseType, Range.getEnd());
+  VarDecl *VD =
+      VarDecl::Create(Ctx, DCT, Range.getEnd(), Range.getEnd(),
+                      VN.getAsIdentifierInfo(), BaseType, TInfo, SC_None);
```
```suggestion
  VarDecl *VD = buildVarDecl(S, Range.getEnd(), BaseType, "_s");
```

jyu2-git replied:
Changed.  Thanks.

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -20796,6 +20796,150 @@ struct MappableVarListInfo {
+  DeclarationNameInfo MapperId;
+  OpenMPDirectiveKind DKind = Stack->getCurrentDirective();
+
+  OMPClause *MapClasue = S.OpenMP().ActOnOpenMPMapClause(
```
```suggestion
  OMPClause *MapClause = S.OpenMP().ActOnOpenMPMapClause(
```

jyu2-git replied:
Changed.  :-(

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -20796,6 +20796,150 @@ struct MappableVarListInfo {
+      /*IsMapTypeImplicit=*/true, SourceLocation(), SourceLocation(), SExprs,
+      OMPVarListLocTy());
+  Maps.push_back(MapClasue);
+  return {MapperVarRef, VD};
```
Why need to return pair here, if MapperVarRef points to the same VD?

jyu2-git replied:
Changed.

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -20796,6 +20796,150 @@ struct MappableVarListInfo {
+
+  // Build impilicit map for mapper
+  SmallVector<OMPClause *, 4> Maps;
+  VarDecl *VD;
+  DeclRefExpr *MapperVarRef;
+  std::tie(MapperVarRef, VD) = buildImplicitMap(S, BaseType, Stack, Maps);
```
```suggestion
  auto [MapperVarRef, VD] = buildImplicitMap(S, BaseType, Stack, Maps);
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -20796,6 +20796,150 @@ struct MappableVarListInfo {
+  llvm::DenseMap<const Type *, Expr *> Visited;
+  SmallVector<std::pair<FieldDecl *, unsigned>, 4> ParentChain(1, {nullptr, 1});
+  while (!Types.empty()) {
+    QualType BaseType;
+    FieldDecl *CurFD;
+    std::tie(BaseType, CurFD) = Types.pop_back_val();
```
```suggestion
    auto [BaseType, CurFD] = Types.pop_back_val();
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -20796,6 +20796,150 @@ struct MappableVarListInfo {
+                             DMD, false, SourceLocation(), BaseType, VK_LValue);
+}
+
+static bool IsImplicitMapperNeeded(Sema &S, DSAStackTy *Stack,
```
```suggestion
static bool isImplicitMapperNeeded(Sema &S, DSAStackTy *Stack,
```

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -20796,6 +20796,150 @@ struct MappableVarListInfo {
+      DefaultMapperId.setName(S.Context.DeclarationNames.getIdentifier(
+          &S.Context.Idents.get("default")));
+      DefaultMapperId.setLoc(E->getExprLoc());
+      ExprResult ER = buildUserDefinedMapperRef(
+          S, Stack->getCurScope(), MapperIdScopeSpec, DefaultMapperId, BaseType,
+          /*UnresolvedMapper=*/nullptr);
+      if (ER.isInvalid())
+        continue;
+      It = Visited.try_emplace(BaseType.getTypePtr(), ER.get()).first;
```
I assume it may create a lot of noise in the AST nodes here, better just to perform normal lookup

jyu2-git replied:
Hi Alexey, thanks for the review!
Question:
Are saying, you don't want me to call buildUserDefinedMapperRef for look up UDM? Or something else?

Thanks.

alexey-bataev replied:
Yes. Can you try to just do a lookup for a mapper? I assume, it should not be very hard to implement instead of building declrefexpr here. AT least, give it a try

jyu2-git replied:
Okay, I am trying.

jyu2-git replied:
Added new function hasUserDefinedMapper.

jyu2-git committed:
Thanks Alexey for the review.  This is to add new function hasUserDefinedMapper.

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -20796,6 +20796,204 @@ struct MappableVarListInfo {
+    return true;
+  // Find the first user-defined mapper with a type derived from the desired
+  // type.
+  if (auto *VD = filterLookupForUDReductionAndMapper<ValueDecl *>(
+          Lookups, [&SemaRef, Type, Loc](ValueDecl *D) -> ValueDecl * {
+            if (!D->isInvalidDecl() &&
+                SemaRef.IsDerivedFrom(Loc, Type, D->getType()) &&
+                !Type.isMoreQualifiedThan(D->getType()))
+              return D;
+            return nullptr;
+          })) {
+    CXXBasePaths Paths(/*FindAmbiguities=*/true, /*RecordPaths=*/true,
+                       /*DetectVirtual=*/false);
+    if (SemaRef.IsDerivedFrom(Loc, Type, VD->getType(), Paths)) {
+      if (!Paths.isAmbiguous(SemaRef.Context.getCanonicalType(
+              VD->getType().getUnqualifiedType()))) {
+        if (SemaRef.CheckBaseClassAccess(
+                Loc, VD->getType(), Type, Paths.front(),
+                /*DiagID=*/0) != Sema::AR_inaccessible) {
+          return true;
+        }
+      }
+    }
+  }
```
Simplify this block by early exiting where possible.

jyu2-git replied:
Thanks.  Changed.

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -20796,6 +20796,204 @@ struct MappableVarListInfo {
+  SmallVector<UnresolvedSet<8>, 4> Lookups;
+  LookupResult Lookup(SemaRef, MapperId, Sema::LookupOMPMapperName);
+  Lookup.suppressDiagnostics();
+  if (S)
+    while (S && SemaRef.LookupParsedName(Lookup, S, &MapperIdScopeSpec,
```
```suggestion
    while (S && SemaRef.LookupParsedName(Lookup, S, &MapperIdScopeSpec,
```

alexey-bataev reviewed (commented):
Add info to OpenMPSupport.rst and release notes

jyu2-git committed:
Thanks Alexey for the review.  This address his' comments.

jyu2-git committed:
Clang format.

jyu2-git committed:
Format.

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -20796,6 +20796,204 @@ struct MappableVarListInfo {
+    if (IsAmbiguous)
+      return false;
+    if (SemaRef.CheckBaseClassAccess(Loc, VD->getType(), Type, Paths.front(),
+                                     /*DiagID=*/0) != Sema::AR_inaccessible) {
+      return true;
+    }
```
Drop braces

alexey-bataev reviewed (approved):
LG with a nit

jyu2-git committed:
Small fix.

jyu2-git commented:
> LG with a nit

Thanks Alexey for the code review!!
